enclave-runner/src/usercalls/mod.rs

```diff
@@ -76,148 +83,307 @@ impl<T> Write for ReadOnly<T> {
+    while let Some(task) = queue.pop_front() {
+        task.notify();
     }
 }
 
-impl<'a> SharedStream<'a> for io::Stderr {
-    type Inner = WriteOnly<io::StderrLock<'a>>;
+impl AsyncStreamAdapter {
+    fn poll_read(&mut self, buf: &mut [u8]) -> tokio::io::Result<Async<usize>> {
+        match self.stream.poll_read(buf) {
+            Ok(Async::NotReady) => {
+                self.read_queue.push_back(tokio::prelude::task::current());
+                Ok(Async::NotReady)
+            }
+            Ok(Async::Ready(ret)) => {
+                notify_other_tasks(&mut self.read_queue);
+                Ok(Async::Ready(ret))
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    fn poll_write(&mut self, buf: &[u8]) -> tokio::io::Result<Async<usize>> {
+        match self.stream.poll_write(buf) {
+            Ok(Async::NotReady) => {
+                self.write_queue.push_back(tokio::prelude::task::current());
+                Ok(Async::NotReady)
+            }
+            Ok(Async::Ready(ret)) => {
+                notify_other_tasks(&mut self.write_queue);
+                Ok(Async::Ready(ret))
+            }
+            Err(e) => Err(e),
+        }
+    }
 
-    fn lock(&'a self) -> Self::Inner {
-        WriteOnly(io::Stderr::lock(self))
+    fn poll_flush(&mut self) -> tokio::io::Result<Async<()>> {
+        match self.stream.poll_flush() {
+            Ok(Async::NotReady) => {
+                self.flush_queue.push_back(tokio::prelude::task::current());
+                Ok(Async::NotReady)
+            }
+            Ok(Async::Ready(ret)) => {
+                notify_other_tasks(&mut self.flush_queue);
+                Ok(Async::Ready(ret))
+            }
+            Err(e) => Err(e),
+        }
     }
 }
 
-impl<S: 'static + Send + Sync> SyncStream for S
-where
-    for<'a> &'a S: Read + Write,
-{
-    fn read(&self, buf: &mut [u8]) -> IoResult<usize> {
-        Read::read(&mut { self }, buf)
+struct AsyncStreamContainer {
+    inner: Lock<AsyncStreamAdapter>,
+}
+
+impl AsyncStreamContainer {
+    fn new<S: AsyncStream>(s: Box<S>) -> Self {
+        AsyncStreamContainer {
+            inner: Lock::new(AsyncStreamAdapter {
+                stream: s,
+                read_queue: VecDeque::new(),
+                write_queue: VecDeque::new(),
+                flush_queue: VecDeque::new(),
+            }),
+        }
     }
 
-    fn write(&self, buf: &[u8]) -> IoResult<usize> {
-        Write::write(&mut { self }, buf)
+    async fn async_read(&self, buf: &mut [u8]) -> IoResult<usize> {
+        tokio::prelude::future::poll_fn(|| match self.inner.clone().poll_lock() {
+            Async::NotReady => Ok(Async::NotReady),
+            Async::Ready(mut adapter) => adapter.poll_read(buf),
+        })
+        .compat()
+        .await
     }
 
-    fn flush(&self) -> IoResult<()> {
-        Write::flush(&mut { self })
+    async fn async_read_alloc(&self) -> IoResult<Vec<u8>> {
+        let mut vec: Vec<u8> = vec![0; 8192];
```

Comment:
Don't default to a vec at this level. Implementors of `SyncStream` used to be able to implement `read_alloc`. We should make sure implementors of `AsyncStream` can do the same.

Reply:
so, as I understand this, we should have a function like `poll_read_alloc` in the trait `AsyncStream` which can then also be overridden by the implementers of AsyncStream.

Reply:
Yes